sway-lib-std/src/bytes.sw

```diff
@@ -0,0 +1,717 @@
+    ///
+    /// let bytes = Bytes::new();
+    /// assert(bytes.is_empty());
+    /// bytes.push(5);
+    /// assert(!bytes.is_empty());
+    /// bytes.clear()
+    /// assert(bytes.is_empty());
+    /// ```
+    pub fn is_empty(self) -> bool {
+        self.len == 0
+    }
+}
+
+// Need to use seperate impl blocks for now: https://github.com/FuelLabs/sway/issues/1548
+impl Bytes {
+    /// Creates a Bytes from a Vec<u8>.
+    ///
+    /// ### Examples
+    ///
+    /// ```sway
+    /// use std:bytes::Bytes;
+    ///
+    /// let mut vec = Vec::new();
+    /// let a = 5u8;
+    /// let b = 7u8;
+    /// let c = 9u8
+    ///
+    /// vec.push(a);
+    /// vec.push(b);
+    /// vec.push(c);
+    ///
+    /// let bytes = Bytes::from_vec_u8(vec);
+    ///
+    /// assert(bytes.len == 3);
+    /// assert(bytes.get(0).unwrap() == a);
+    /// assert(bytes.get(1).unwrap() == b);
+    /// assert(bytes.get(2).unwrap() == c);
+    /// ```
+    pub fn from_vec_u8(ref mut vec: Vec<u8>) -> Self {
+        let mut bytes = Bytes::new();
+        let mut i = 0;
+        let length = vec.len();
+        while i < length {
+            bytes.push(vec.get(i).unwrap());
+            i += 1;
+        };
+        bytes
+    }
+
+    /// Creates a Vec<u8> from a Bytes.
+    ///
+    /// ### Examples
+    ///
+    /// ```sway
+    /// use std:bytes::Bytes;
+    ///
+    /// let mut bytes = Bytes::new();
+    /// let a = 5u8;
+    /// let b = 7u8;
+    /// let c = 9u8
+    /// bytes.push(a);
+    /// bytes.push(b);
+    /// bytes.push(c);
+    ///
+    /// assert(bytes.len() == 3);
+    ///
+    /// let vec = bytes.into_vec_u8();
+    ///
+    /// assert(vec.len() == 3);
+    /// assert(vec.get(0).unwrap() == a);
+    /// assert(vec.get(1).unwrap() == b);
+    /// assert(vec.get(2).unwrap() == c);
+    /// ```
+    pub fn into_vec_u8(self) -> Vec<u8> {
+        let mut vec = Vec::new();
+        let mut i = 0;
+        let length = self.len;
+        while i < length {
+            vec.push(self.get(i).unwrap());
+            i += 1;
+        };
+        vec
+    }
+}
+
+////////////////////////////////////////////////////////////////////
+// Tests
+////////////////////////////////////////////////////////////////////
+fn setup() -> (Bytes, u8, u8, u8) {
+    let mut bytes = Bytes::new();
+    let a = 5u8;
+    let b = 7u8;
+    let c = 9u8;
+    bytes.push(a);
+    bytes.push(b);
+    bytes.push(c);
+    (bytes, a, b, c)
+}
+
+#[test()]
+fn test_new_bytes() {
+    let bytes = Bytes::new();
+    assert(bytes.len() == 0);
+}
+#[test()]
+fn test_push() {
+    let (_, a, b, c) = setup();
+    let mut bytes = Bytes::new();
+    bytes.push(a);
+    assert(bytes.len() == 1);
+    bytes.push(b);
+    assert(bytes.len() == 2);
+    bytes.push(c);
+    assert(bytes.len() == 3);
+}
+#[test()]
+fn test_pop() {
+    let (mut bytes, a, b, c) = setup();
+    assert(bytes.len() == 3);
+
+    let first = bytes.pop();
+
+    assert(first.unwrap() == c);
+    assert(bytes.len() == 2);
+}
```

Comment:
@otrho This tests that we can push and pop bytes. 
`push` is using that `Raw_ptr::write_byte()` under the hood.

Reply:
But you don't have a test which is pushing, then popping and confirming the value is correct.  I only saw a test which pushes and then checks the len has incremented..?

Reply:
Ah. the `setup` function in the test pushes 3 values.
```
fn setup() -> (Bytes, u8, u8, u8) {
    let mut bytes = Bytes::new();
    let a = 5u8;
    let b = 7u8;
    let c = 9u8;
    bytes.push(a);
    bytes.push(b);
    bytes.push(c);
    (bytes, a, b, c)
}
```

Reply:
Oh, OK.  I still don't think it should work, but I'll take a look at the final asm and/or reassess how it works in my head. :slightly_smiling_face:

Reply:
Update: I've tried to improve the testing for this; the new test for pop is:
```
#[test()]
fn test_pop() {
    let (mut bytes, a, b, c) = setup();
    assert(bytes.len() == 3);
    bytes.push(42u8);
    bytes.push(11u8);
    bytes.push(69u8);
    bytes.push(100u8);
    bytes.push(200u8);
    bytes.push(255u8);
    bytes.push(180u8);
    bytes.push(17u8);
    bytes.push(19u8);
    assert(bytes.len() == 12);

    let first = bytes.pop();
    assert(first.unwrap() == 19u8);
    assert(bytes.len() == 11);

    let second = bytes.pop();
    assert(second.unwrap() == 17u8);
    assert(bytes.len() == 10);

    let third = bytes.pop();
    assert(third.unwrap() == 180u8);
    assert(bytes.len() == 9);

    bytes.pop();
    bytes.pop();
    bytes.pop();
    bytes.pop();
    bytes.pop();
    bytes.pop();
    assert(bytes.len() == 3);

    assert(bytes.pop().unwrap() == c);
    assert(bytes.pop().unwrap() == b);
    assert(bytes.pop().unwrap() == a);

    assert(bytes.pop().is_none() == true);
    assert(bytes.len() == 0);
}
```

Reply:
OK, great!  I need to take a closer look and work out why I was mistaken. :+1:

Reply:
Hrmm. If I modify `fn write_byte` to look like this, it doesn't appear to break any tests:
```
pub fn write_byte(self, val: u8) {
        asm(ptr: self, val: val) {
            sb ptr val i0;
        };
    }
```